clients/src/main/java/org/apache/kafka/clients/producer/internals/Sender.java

```diff
@@ -452,11 +452,12 @@ private boolean maybeSendTransactionalRequest(long now) {
                 if (targetNode != null) {
                     if (nextRequestHandler.isRetry())
                         time.sleep(nextRequestHandler.retryBackoffMs());
+                    long currentTimeMs = time.milliseconds();
                     ClientRequest clientRequest = client.newClientRequest(
-                        targetNode.idString(), requestBuilder, now, true, requestTimeoutMs, nextRequestHandler);
-                    transactionManager.setInFlightTransactionalRequestCorrelationId(clientRequest.correlationId());
+                        targetNode.idString(), requestBuilder, currentTimeMs, true, requestTimeoutMs, nextRequestHandler);
                     log.debug("Sending transactional request {} to node {}", requestBuilder, targetNode);
-                    client.send(clientRequest, now);
+                    client.send(clientRequest, currentTimeMs);
```

Comment:
It seems to me that this is the main fix. Do we really want to redefine 'now' in this method as you do on L455? It is used in multiple places in `run` and I wonder if there will be side effects of using a different value here.

Reply:
Hmm.. We definitely have to update the time after the blocking operations above, but it's a fair point that the old `now` continues to be used in `run` afterwards. We actually have the same problem in `maybeWaitForProducerId`, which is also a blocking operation. I don't see a way around it except to change the logic in `run` to treat `now` as more of a `startTimeMs`.

Reply:
You're right that accounting for the blocking operations is key. 

Initially, I thought the main issue with recomputing the `currentTime` in every down stream call is that it will likely affect the validity of a bunch of the test cases in `SenderTest`. But I looked at the cases they should be OK with such a change. 

I think standardizing on treating `now` as more of a `startTimeMs` makes sense.

Reply:
Thanks for reviewing. I ended up changing the signature so that it removes the timing assumption. Literally every call to this method was also `run(time.milliseconds())`, so I got rid of the argument and renamed to `runOnce`.

Reply:
I like the refactoring too: we had similar pattern of passing `now` in Streamthread loop but later I refactored it to `advanceNowAndComputeLatency` whenever a blocking call is made for the same reason.